tests/virt_autotest/qem/live_migration_src.pm

```diff
@@ -0,0 +1,217 @@
+        $vm_info->{ip} = script_output("cat $images_path/guestip/$vm_name");
+        if ("$cmd" eq "xl migrate") {
+            record_info("$cmd $vm_name $dst_ip");
+            assert_script_run("$cmd $vm_name $dst_ip");
+        } else {
+            record_info("$cmd $vm_name $vm_info->{dsturi}");
+            assert_script_run("$cmd $vm_name $vm_info->{dsturi}");
+        }
+        # visit web server running in vm check it's content
+        validate_script_output("curl -s http://$vm_info->{ip}/guest", sub { m/This is guest $vm_name/ });
+        #test vm disk write and read
+        assert_script_run("echo $vm_name > /tmp/test_vm_write");
+        assert_script_run("curl -s -k -T /tmp/test_vm_write -u bernhard:$testapi::password scp://$vm_info->{ip}/~/test_vm_write");
+        validate_script_output("curl -s -k -u bernhard:$testapi::password scp://$vm_info->{ip}/~/test_vm_write", sub { m/$vm_name/ });
+        assert_script_run("rm $images_path/guestip/$vm_name");
+        #shutdown migrated vm
+        if ($cmd eq "xl migrate") {
+            assert_script_run("ssh root\@$dst_ip xl shutdown $vm_name");
+        } else {
+            assert_script_run("ssh root\@$dst_ip virsh shutdown $vm_name");
+        }
+        #wait for shutdown complete
+        sleep 15;
+        #start again for next migration test
+        assert_script_run("virsh start $vm_name") if ($cmd ne $vm_info->{virsh_cmds}->[-1]);
+    }
+}
+
+sub cleanup {
+    my ($vm_name, $vm_info) = @_;
+    assert_script_run("virsh dumpxml $vm_name > $images_path/vm-configs/$vm_name.xml") unless ($vm_info->{reuse});
+    assert_script_run("virsh undefine $vm_name");
+}
+
+sub run {
+    select_console 'root-ssh';
+    $dst_ip = get_job_autoinst_vars(get_parents->[0])->{SUT_IP};
+    record_info("DST: $dst_ip");
+    record_info("SUT: $sut_ip");
+    assert_script_run("mount $sut_ip:/tmp/vm_images $images_path") unless get_var("NFS_PATH");
+    assert_script_run("rm -rf $images_path/guestip; mkdir $images_path/guestip");
+    assert_script_run("mkdir -p $images_path/vm-configs");
+    my $guests = get_var_array('GUESTS');
+    #start all guests
+    for my $guest (@{$guests}) { start_guest($guest, $vms); }
+
+    assert_script_run('if [ -d /root/.ssh ]; then rm -rf /root/.ssh; fi');
+    assert_script_run('ssh-keygen -t rsa -q -f "/root/.ssh/id_rsa" -N ""');
+
+    mutex_wait 'dst_ready';
+
+    assert_script_run("ssh-keyscan $dst_ip > /root/.ssh/known_hosts");
+    assert_script_run("curl -s -k --user root:$testapi::password -T .ssh/id_rsa.pub sftp://$dst_ip/root/.ssh/authorized_keys  --ftp-create-dirs");
+    assert_script_run("ssh root\@$dst_ip mount $sut_ip:/tmp/vm_images $images_path") unless get_var("NFS_PATH");
+
+    my $n         = keys %$vms;
+    my $time_cont = 0;
+    #Check each vm if their have got IP address. If they have, their install or stratup are supposed to complete and ready for migration test.
+    until ($n == 0) {
```

Comment:
This loop mixes functions  "waiting guest install finish", "migration", etc. Maybe it is very efficient(without time waste), but it will be very challenging to make the code robust and being able to handle various situations.

Can we sacrifice some efficiency and let the code easy, clean and robust? For example,  split them like this:
1) let all guest preparation finish: 
either fresh install or reuse, and high chance to be either case when multiple vm applied, so efficiency lose is not high most of the time
exclude failure vms, and continue successful vms
2) vm migration one by one
3) cleanup for env/vm

Reply:
Here is a user scenario:  a costumer has  host where some VMs are running.  Some VMs have web servers running in them. Some VMs are being installed.  The others are being live migrated one after another by customer. 
This is the test scenario of this automation.

What you suggested is actually my previous version of code.  In production, costumers rarely do migration that way.

Reply:
> Can we sacrifice some efficiency and let the code easy, clean and robust? For example, split them like this:
> 
> 1. let all guest preparation finish:
>    either fresh install or reuse, and high chance to be either case when multiple vm applied, so efficiency lose is not high most of the time
>    exclude failure vms, and continue successful vms
> 2. vm migration one by one
> 3. cleanup for env/vm

This will add two more loops. 2, 3.  They are unnecessary and make code unclean and unrobust.

Reply:
> This loop mixes functions "waiting guest install finish", "migration", etc. Maybe it is very efficient(without time waste), but it will be very challenging to make the code robust and being able to handle various situations.
This loop is just to check if a VM is ready for migration test.  No matter if it's just installed or reused, they are the same from migration point of view.

Reply:
> This loop mixes functions "waiting guest install finish", "migration", etc. Maybe it is very efficient(without time waste), but it will be very challenging to make the code robust and being able to handle various situations.
The code is robust, I tested a lot and could not break it now :).
If it tends to trigger VT bugs it's just what we want. 
Do you have any concern?